mayham-game/Goal.gd

```diff
@@ -4,11 +4,48 @@ extends Area2D
 # var a = 2
 # var b = "textvar"
 
+var player_list = preload("res://PlayerLinkedList.gd")
+
+var players_in_goal = null
+var players_not_in_goal = null
+
+func init(players):
+	players_in_goal = player_list.new()
+	players_not_in_goal = player_list.new()
+	for player in players:
+		players_not_in_goal.push_back(player)
+
 func _ready():
+	connect("body_entered", self, "_on_Goal_entered")
+	connect("body_exited", self, "_on_Goal_exited")
+
 	# Called when the node is added to the scene for the first time.
 	# Initialization here
 	pass
 
+func _on_Goal_entered(body):
+	if body.is_class("StaticBody2D"):
+		return
+	var player = players_not_in_goal.pop_by_number(body.get_number())
+	players_in_goal.push_back(player)
+
+func _on_Goal_exited(body):
+	var player = players_in_goal.pop_by_number(body.get_number())
+	players_not_in_goal.push_back(player)
+
+func _on_Goal_moved():
+	players_not_in_goal.push_all(players_in_goal.pop_all())
```

Comment:
Maybe not in this merge since this hasn't been used yet from what I can see, but we will need to think about the case where the goal's new position is already intersecting the player.

Reply:
Yes that is a good point.

Reply:
We can do this by calling get_overlapping_bodies() which returns an array of all the intersecting PhysicsBody2D objects.

Reply:
Looks good